airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,482 @@
+        self._matchers = matchers
+
+    def matches(self, request: Any) -> bool:
+        """Tell if this policy matches specific request and should apply to it
+
+        :param request:
+        :return: True if policy should apply to this request, False - otherwise
+        """
+
+        if not self._matchers:
+            return True
+        return any(matcher(request) for matcher in self._matchers)
+
+
+class FixedWindowCallRatePolicy(BaseCallRatePolicy):
+    def __init__(self, window: TimeWindow, call_limit: int, **kwargs: Any):
+        """A policy that allows {call_limit} calls within a {window} time interval
+
+        :param window:
+        :param call_limit:
+        """
+
+        self._current_window = TimeWindow(window.start, window.end)
+        self._offset = window.end - window.start
+        self._call_limit = call_limit
+        self._calls_num = 0
+        super().__init__(**kwargs)
+
+    def try_acquire(self, request: Any, weight: int) -> None:
+        if not self.matches(request):
+            raise ValueError("Request does not match the policy")
+
+        self._update_current_window()
+
+        if self._calls_num >= self._call_limit:
+            reset_in = self._current_window.end - datetime.datetime.now()
+            error_message = (
+                f"reached maximum number of allowed calls {self._call_limit} " f"per {self._offset} interval, next reset in {reset_in}."
+            )
+            raise CallRateLimitHit(
+                error=error_message,
+                item=request,
+                weight=weight,
+                rate=f"{self._call_limit} per {self._offset}",
+                time_to_wait=reset_in,
+            )
+
+        self._calls_num += 1
+
+    def update(self, available_calls: Optional[int], call_reset_ts: Optional[datetime.datetime]) -> None:
+        """Update call rate counters, by default only reacts to decreasing updates of available_calls and changes to call_reset_ts
+
+        :param available_calls:
+        :param call_reset_ts:
+        """
+        self._update_current_window()
+        current_available_calls = self._call_limit - self._calls_num
+
+        if available_calls is not None and current_available_calls > available_calls:
+            logger.info("got update from api, adjusting available calls from %s to %s", current_available_calls, available_calls)
+            self._calls_num = self._call_limit - available_calls
+
+        if call_reset_ts is not None and call_reset_ts != self._current_window.end:
+            logger.info("got update from api, adjusting reset time from %s to %s", self._current_window.end, call_reset_ts)
+            self._current_window.end = call_reset_ts
+
+    def _update_current_window(self) -> None:
+        now = datetime.datetime.now()
+        if now > self._current_window.end:
+            logger.info("current window %s has being reset, %s calls available now", self._current_window, self._call_limit)
```

Comment:
we should probably log this as debug (and confirmed it is logged when running with the `--debug` flag because this will result in one line per second which can be massive on long syncs